File: code-awesome/node-express-react/voting_app/backend/src/components/database-layer.js
// database-layer.js
// This file represents the backend layer for interacting with the database.
// It's responsible for handling data retrieval and storage.
// It includes a vulnerability demonstration for educational purposes only.

const mysql = require('mysql');

// Configuration for the MySQL database
const dbConfig = {
    host: 'localhost',
    user: 'root',
    password: 'your_password', // Replace with your actual password
    database: 'voting_app'
};

// Create a connection to the database
const db = mysql.createConnection(dbConfig);

// Event listener for database connection events
db.connect(err => {
    if (err) {
        console.error('Error connecting to database:', err);
    } else {
        console.log('Connected to database successfully!');
    }
});

// Function to add a vote
function addVote(voteData) {
    const query = 'INSERT INTO votes (user_id, candidate_id) VALUES (?, ?)';
    db.query(query, [voteData.user_id, voteData.candidate_id], err => {
        if (err) {
            console.error('Error inserting vote:', err);
        } else {
            console.log('Vote added successfully!');
        }
    });
}

// Function to retrieve all votes
function getAllVotes() {
    const query = 'SELECT * FROM votes';
    db.query(query, (err, results) => {
        if (err) {
            console.error('Error retrieving votes:', err);
            return null;
        }
        console.log('All Votes:', results);
        return results;
    });
}

// Function to retrieve votes for a specific candidate
function getVotesForCandidate(candidateId) {
  const query = 'SELECT * FROM votes WHERE candidate_id = ?';
  db.query(query, [candidateId], (err, results) => {
      if (err) {
          console.error('Error retrieving votes for candidate:', err);
          return null;
      }
      console.log('Votes for Candidate ' + candidateId + ':', results);
      return results;
  });
}

// Export the functions to be used by other modules
module.exports = {
    addVote,
    getAllVotes,
    getVotesForCandidate
};

File: code-awesome/ruby-on-rails-react/voting_app/backend/src/components/database-layer.rb
# database-layer.rb

# This file represents the database layer of the voting application.
# It handles interactions with the PostgreSQL database.

# Constants
DATABASE_URL = "postgres://voting_user:voting_password@localhost:5432/voting_app"
# Variables
votes_table_name = 'votes'
user_table_name = 'users'
# Helper functions

def connect_to_database
  # Simulate database connection logic.  In a real application, this would
  # use a library like pg or ActiveRecord.
  puts "Connecting to database..."
  # Placeholder for actual database connection logic.
  # In a real application, you would use a database library to
  # establish a connection to the database.
  # For example:
  # db = PG.new(DATABASE_URL)
  # begin
  #   db.execute("SELECT version()")
  # end
  puts "Connected successfully!"
  return true # Simulate successful connection.
end

def add_vote(user_id, candidate_id)
  # Simulates adding a vote to the database.
  # In a real application, this would execute a SQL query to insert a
  # new row into the 'votes' table.
  puts "Adding vote for user #{user_id} to candidate #{candidate_id}..."
  # Simulate database insertion
  # db.execute("INSERT INTO votes (user_id, candidate_id) VALUES (#{user_id}, #{candidate_id})")
  puts "Vote added successfully!"
  return true
end

def get_votes_for_candidate(candidate_id)
  # Simulates retrieving votes for a candidate from the database.
  # In a real application, this would execute a SQL query to select
  # all rows from the 'votes' table where the candidate_id matches the
  # provided candidate_id.
  puts "Retrieving votes for candidate #{candidate_id}..."
  # Simulate database retrieval
  # results = db.execute("SELECT * FROM votes WHERE candidate_id = #{candidate_id}")
  # return results
  puts "Votes retrieved successfully!"
  return [{user_id: 1, candidate_id: 1}, {user_id: 2, candidate_id: 2}]
end

def get_all_votes()
  # Simulates getting all votes from the database.
  puts "Getting all votes..."
  # Simulate database retrieval
  # results = db.execute("SELECT * FROM votes")
  # return results
  puts "Votes retrieved successfully!"
  return [{user_id: 1, candidate_id: 1}, {user_id: 2, candidate_id: 2}, {user_id: 3, candidate_id: 1}]
end
# Example Usage (for testing)
# if __FILE__ == '__main__'
#   # Simulate database connection.
#   if !connect_to_database()
#     puts "Failed to connect to the database."
#     exit
#   end

#   # Add some votes.
#   add_vote(1, 1)
#   add_vote(2, 2)

#   # Get votes for a candidate.
#   votes = get_votes_for_candidate(1)
#   puts "Votes for candidate 1: #{votes}"

#   all_votes = get_all_votes()
#   puts "All votes: #{all_votes}"
# end